src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentCodeActionResolver.cs

```diff
@@ -0,0 +1,151 @@
+        {
+            return null;
+        }
+
+        var path = FilePathNormalizer.Normalize(actionParams.Uri.GetAbsoluteOrUNCPath());
+        var directoryName = Path.GetDirectoryName(path).AssumeNotNull();
+        var templatePath = Path.Combine(directoryName, "Component.razor");
+        var componentPath = FileUtilities.GenerateUniquePath(templatePath, ".razor");
+        var componentName = Path.GetFileNameWithoutExtension(componentPath);
+
+        // VS Code in Windows expects path to start with '/'
+        componentPath = _languageServerFeatureOptions.ReturnCodeActionAndRenamePathsWithPrefixedSlash && !componentPath.StartsWith('/')
+            ? '/' + componentPath
+            : componentPath;
+
+        var newComponentUri = new UriBuilder
+        {
+            Scheme = Uri.UriSchemeFile,
+            Path = componentPath,
+            Host = string.Empty,
+        }.Uri;
+
+        using var _ = StringBuilderPool.GetPooledObject(out var builder);
+
+        var syntaxTree = componentDocument.GetSyntaxTree();
+        foreach (var usingDirective in GetUsingsInDocument(syntaxTree))
+        {
+            builder.AppendLine(usingDirective);
+        }
+
+        var extractedText = text.GetSubTextString(new TextSpan(actionParams.Start, actionParams.End - actionParams.Start)).Trim();
+        builder.Append(extractedText);
+
+        var start = componentDocument.Source.Text.Lines.GetLinePosition(actionParams.Start);
+        var end = componentDocument.Source.Text.Lines.GetLinePosition(actionParams.End);
+        var removeRange = new Range
+        {
+            Start = new Position(start.Line, start.Character),
+            End = new Position(end.Line, end.Character)
+        };
+
+        var documentChanges = new SumType<TextDocumentEdit, CreateFile, RenameFile, DeleteFile>[]
+        {
+            new CreateFile { Uri = newComponentUri },
+            new TextDocumentEdit
+            {
+                TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = actionParams.Uri },
+                Edits =
+                [
+                    new TextEdit
+                    {
+                        NewText = $"<{componentName} />",
+                        Range = removeRange,
+                    }
+                ],
+            },
+            new TextDocumentEdit
+            {
+                TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = newComponentUri },
+                Edits  =
+                [
+                    new TextEdit
+                    {
+                        NewText = builder.ToString(),
+                        Range = new Range { Start = new Position(0, 0), End = new Position(0, 0) },
+                    }
+                ],
+            }
+        };
+
+        return new WorkspaceEdit
+        {
+            DocumentChanges = documentChanges,
+        };
+    }
+
+    private static IEnumerable<string> GetUsingsInDocument(RazorSyntaxTree syntaxTree)
```

Comment:
Is there an issue to have this just include the minimum set?

Reply:
no, I'll create one. There was code to do that but I am not sure it was right and there were no tests so I removed for now. 

https://github.com/dotnet/razor/issues/11025